src/XMakeTasks/AssemblyDependency/AssemblyInformation.cs

```diff
@@ -592,4 +599,170 @@ private void FreeAsmMeta(IntPtr asmMetaPtr)
+                    const uint OptionalPEHeaderSize = 224;
+                    const uint OptionalPEPlusHeaderSize = 240;
+                    const uint SectionHeaderSize = 40;
+
+                    // There must be room for at least the PE header offset and size, and sections
+                    if (sr.BaseStream.Length < PEHeaderPointerOffset + 4 + PEHeaderSize + OptionalPEHeaderSize + SectionHeaderSize)
+                        return string.Empty;
+
+                    // PE header
+                    sr.BaseStream.Position = PEHeaderPointerOffset;
+                    var peHeaderOffset = sr.ReadUInt32();
+
+                    if (peHeaderOffset + 4 + PEHeaderSize + OptionalPEHeaderSize + SectionHeaderSize >= sr.BaseStream.Length)
+                        return string.Empty;
+
+                    sr.BaseStream.Position = peHeaderOffset;
+                    if (!ReadBytes(sr, (byte)'P', (byte)'E', 0, 0)) // PE header signature
+                        return string.Empty;
+
+                    peHeaderOffset += 4; // Skip signature
+
+                    // Read the number of sections
+                    sr.BaseStream.Position = peHeaderOffset + 2;
+                    var numberOfSections = sr.ReadUInt16();
+                    if (numberOfSections > 96)
+                        return string.Empty; // There can't be more than 96 sections, something is wrong
+
+                    // Optional header start
+                    var optionalHeaderOffset = peHeaderOffset + PEHeaderSize;
+
+                    uint cliHeaderRvaOffset;
+                    uint optionalPEHeaderSize;
+
+                    // Get the offset to the CLI header RVA. That's in the optional header,
+                    // which can have two formats, PE32 or PE32+
+                    sr.BaseStream.Position = optionalHeaderOffset;
+                    var magicNumber = sr.ReadUInt16();
+
+                    if (magicNumber == 0x10b) // PE32
+                    {
+                        optionalPEHeaderSize = OptionalPEHeaderSize;
+                        cliHeaderRvaOffset = optionalHeaderOffset + 208;
+                    }
+                    else if (magicNumber == 0x20b) // PE32+
+                    {
+                        optionalPEHeaderSize = OptionalPEPlusHeaderSize;
+                        cliHeaderRvaOffset = optionalHeaderOffset + 224;
+                    }
+                    else
+                        return string.Empty;
+
+                    // Read the CLI header RVA
+                    sr.BaseStream.Position = cliHeaderRvaOffset;
+                    var cliHeaderRva = sr.ReadUInt32();
+                    if (cliHeaderRva == 0)
+                        return string.Empty; // No CLI section
+
+                    var sectionOffset = optionalHeaderOffset + optionalPEHeaderSize;
+
+                    var sections = new HeaderInfo [numberOfSections];
+                    for (int n = 0; n < numberOfSections; n++)
```

Comment:
In the beginning I see a lot of careful checking for the length of the stream? In this case looping over max 96 sections is not bounded. Did I perhaps miss a length check?